src/lem/circuit.rs

```diff
@@ -552,26 +646,194 @@ impl Func {
+                        let one = g.global_allocator.get_or_alloc_const(cs, F::ONE)?;
+
+                        let divisor = pick(
+                            &mut cs.namespace(|| "maybe-dummy divisor"),
+                            b_is_zero,
+                            &one,
+                            b_num,
+                        )?;
+
+                        let quotient = div(&mut cs.namespace(|| "quotient"), a_num, &divisor)?;
+
+                        let tag = g
+                            .global_allocator
+                            .get_or_alloc_const(cs, Tag::Expr(Num).to_field())?;
+                        let c = AllocatedPtr::from_parts(tag, quotient);
+                        bound_allocations.insert(tgt.clone(), c);
+                    }
+                    Op::Lt(tgt, a, b) => {
+                        let a = bound_allocations.get(a)?;
+                        let b = bound_allocations.get(b)?;
+                        let tag = g
+                            .global_allocator
+                            .get_or_alloc_const(cs, Tag::Expr(Num).to_field())?;
+                        let (preallocated_preimg, lt) =
+                            &g.preallocated_less_than_slots[next_slot.consume_less_than()];
+                        for (i, n) in [a.hash(), b.hash()].into_iter().enumerate() {
+                            implies_equal(
+                                &mut cs.namespace(|| {
+                                    format!("implies equal for component {i} (OP {:?})", &op)
+                                }),
+                                not_dummy,
+                                n,
+                                &preallocated_preimg[i],
+                            )?;
+                        }
+                        let c = AllocatedPtr::from_parts(tag, lt.clone());
+                        bound_allocations.insert(tgt.clone(), c);
+                    }
+                    Op::BitAnd(tgt, a, b) => {
+                        let a = bound_allocations.get(a)?;
+                        let a_bits = a.hash().to_bits_le(&mut cs.namespace(|| "bitwise_and"))?;
+                        let mut trunc_bits = Vec::with_capacity(64);
+                        let mut b_rest = *b;
+                        for a_bit in a_bits {
+                            let b_bit = b_rest & 1;
+                            if b_bit == 1 {
+                                trunc_bits.push(a_bit.clone());
+                            } else {
+                                trunc_bits.push(Boolean::Constant(false))
+                            }
+                            b_rest /= 2;
+                        }
+                        let trunc = AllocatedNum::alloc(cs.namespace(|| "trunc"), || {
+                            let val = a
+                                .hash()
+                                .get_value()
+                                .map(|a| F::from_u64(a.to_u64_unchecked() & b))
```

Comment:
This means that `BitAnd` will fail if the input value is larger than U64, right?

Reply:
Meaning: this implementation will produce witnesses that don't satisfy the circuit.

Reply:
No, we only do `BitAnd` on a field element together with a static u64 (or 64 bits), defined by the LEM code. The input can be any field element and the output will always be at most 64 bits. This shouldn't fail on any input